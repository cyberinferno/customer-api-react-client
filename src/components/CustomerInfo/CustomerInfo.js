/**
 * Customer info component
 */

import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import { fetchCustomerInfoStarted } from '../../actions';

class CustomerInfo extends Component {
  componentDidMount() {
    const { id } = this.props;
    this.props.fetchCustomerInfoStarted(id);
  }

  renderCustomerInfo() {
    const { data } = this.props;
    let content;
    if (typeof data.errors !== 'undefined') {
      content = <b>{data.errors[0].msg}</b>;
    } else {
      content = JSON.stringify(data);
      const keyMap = {
        id: 'ID',
        name: 'Name',
        phone: 'Phone Number',
        address: 'Address',
        country: 'Country',
        created_at: 'Created At',
        updated_at: 'Updated At',
      };
      const rows = [];
      Object.keys(data).forEach((key) => {
        rows.push(
          <tr key={key}>
            <th>{keyMap[key] ? keyMap[key] : key}</th>
            <td>{data[key]}</td>
          </tr>
        );
      });
      content = <table className="table table-bordered"><tbody>{rows}</tbody></table>;
    }
    return (
      <div>
        <div className="row">
          <div className="col-sm-8">
            {content}
          </div>
        </div>
        <div className="row" style={{ marginTop: '20px' }}>
          <div className="col-sm-12">
            <Link to="/" className="btn btn-danger">Back</Link>
          </div>
        </div>
      </div>
    );
  }

  render() {
    let content;
    const { loading, error } = this.props;
    if (loading) {
      content = <b>Loading...</b>;
    } else {
      if (error !== '') {
        content = <b>error</b>;
      } else {
        content = this.renderCustomerInfo();
      }
    }
    return (
      <div>
        {content}
      </div>
    );
  }
}

CustomerInfo.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string.isRequired,
  fetchCustomerInfoStarted: PropTypes.func.isRequired,
  data: PropTypes.object.isRequired,
  id: PropTypes.string,
};

const mapStateToProps = (state) => {
  const {
    loading,
    data,
    error,
  } = state.customerInfo;

  return {
    loading,
    data,
    error,
  };
};

export default connect(
  mapStateToProps,
  {
    fetchCustomerInfoStarted,
  }
)(CustomerInfo);
