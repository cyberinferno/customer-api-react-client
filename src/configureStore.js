/**
 * Create the store with dynamic reducers
 */

import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import logger from 'redux-logger';

import createReducer from './reducers';
import rootSaga from './sagas';

export default function configureStore(initialState = {}) {
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore(
    createReducer(),
    initialState,
    applyMiddleware(logger, sagaMiddleware)
  );

  sagaMiddleware.run(rootSaga);

  return store;
}
