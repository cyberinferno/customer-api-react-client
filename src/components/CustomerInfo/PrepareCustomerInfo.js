/**
 * Returns customer info component by inserting id param
 */

import React from 'react';
import { useParams } from 'react-router-dom';
import CustomerInfo from './CustomerInfo';

function PrepareCustomerInfo() {
  const { id } = useParams();
  return (<CustomerInfo id={id} />);
}

export default PrepareCustomerInfo;
