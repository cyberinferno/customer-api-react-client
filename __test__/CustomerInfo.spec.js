/**
 * Customer info component test
 */

import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { BrowserRouter as Router } from 'react-router-dom';
import renderer from 'react-test-renderer';
import CustomerInfo from '../src/components/CustomerInfo/CustomerInfo';

configure({ adapter: new Adapter() });

describe('>>> Customer info test', () => {
  const initialState = {
    customerInfo: {
      error: '',
      loading: false,
      data: {
        id: 1,
        name: 'Customer 1',
        address: '#1 Nice Apartments, 5th cross, Bengaluru 560001',
        country: 'India',
        phone: 9876543210,
        created_at: '2020-06-25T00:00:00.000Z',
        updated_at: '2020-06-25T00:00:00.000Z',
      },
    },
  };
  const mockStore = configureStore();
  let store;
  let wrapper;

  beforeEach(() => {
    store = mockStore(initialState);
    wrapper = mount(
      <Provider store={store}>
        <Router>
          <CustomerInfo id="1" />
        </Router>
      </Provider>
    );
  });

  it('+++ Component gets rendered', () => {
    expect(wrapper.find(CustomerInfo).length).toEqual(1);
  });

  it('+++ Props match with initialState', () => {
    expect(wrapper.find(CustomerInfo).prop('error')).toEqual(initialState.error);
    expect(wrapper.find(CustomerInfo).prop('loading')).toEqual(initialState.loading);
    expect(wrapper.find(CustomerInfo).prop('data')).toEqual(initialState.data);
  });

  it('+++ Contains table', () => {
    expect(wrapper.find('table').length).toEqual(1);
  });

  it('+++ Contains proper amount of rows', () => {
    expect(wrapper.find('tr').length).toEqual(7);
  });

  it('+++ Table contains the right data', () => {
    expect(wrapper.find('th').get(0).props.children).toBe('ID');
    expect(wrapper.find('th').get(1).props.children).toBe('Name');
    expect(wrapper.find('th').get(2).props.children).toBe('Address');
    expect(wrapper.find('th').get(3).props.children).toBe('Country');
    expect(wrapper.find('th').get(4).props.children).toBe('Phone Number');
    expect(wrapper.find('th').get(5).props.children).toBe('Created At');
    expect(wrapper.find('th').get(6).props.children).toBe('Updated At');
    expect(wrapper.find('td').get(0).props.children).toBe(1);
    expect(wrapper.find('td').get(1).props.children).toBe('Customer 1');
    expect(wrapper.find('td').get(2).props.children).toBe('#1 Nice Apartments, 5th cross, Bengaluru 560001');
    expect(wrapper.find('td').get(3).props.children).toBe('India');
    expect(wrapper.find('td').get(4).props.children).toBe(9876543210);
    expect(wrapper.find('td').get(5).props.children).toBe('2020-06-25T00:00:00.000Z');
    expect(wrapper.find('td').get(6).props.children).toBe('2020-06-25T00:00:00.000Z');
  });

  it('+++ Capturing Snapshot of CustomerInfo', () => {
    const renderedValue = renderer.create(
      <Provider store={store}>
        <Router>
          <CustomerInfo id="1" />
        </Router>
      </Provider>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
