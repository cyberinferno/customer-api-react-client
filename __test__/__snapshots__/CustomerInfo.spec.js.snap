// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Customer info test +++ Capturing Snapshot of CustomerInfo 1`] = `
<div>
  <div>
    <div
      className="row"
    >
      <div
        className="col-sm-8"
      >
        <table
          className="table table-bordered"
        >
          <tbody>
            <tr>
              <th>
                ID
              </th>
              <td>
                1
              </td>
            </tr>
            <tr>
              <th>
                Name
              </th>
              <td>
                Customer 1
              </td>
            </tr>
            <tr>
              <th>
                Address
              </th>
              <td>
                #1 Nice Apartments, 5th cross, Bengaluru 560001
              </td>
            </tr>
            <tr>
              <th>
                Country
              </th>
              <td>
                India
              </td>
            </tr>
            <tr>
              <th>
                Phone Number
              </th>
              <td>
                9876543210
              </td>
            </tr>
            <tr>
              <th>
                Created At
              </th>
              <td>
                2020-06-25T00:00:00.000Z
              </td>
            </tr>
            <tr>
              <th>
                Updated At
              </th>
              <td>
                2020-06-25T00:00:00.000Z
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="row"
      style={
        Object {
          "marginTop": "20px",
        }
      }
    >
      <div
        className="col-sm-12"
      >
        <a
          className="btn btn-danger"
          href="/"
          onClick={[Function]}
        >
          Back
        </a>
      </div>
    </div>
  </div>
</div>
`;
